layer/layer_data.h

```diff
@@ -62,67 +144,91 @@ class LayerData {
     }
   }
 
-  // Records the dispatch table that is associated with |instance|.
+  // Records the dispatch table and instance key that is associated with
+  // |instance|.
   bool AddInstance(VkInstance instance,
                    const VkLayerInstanceDispatchTable& dispatch_table) {
+    InstanceKey key(instance);
+
     absl::MutexLock lock(&instance_dispatch_lock_);
-    return instance_dispatch_map_.insert({instance, dispatch_table}).second;
+    const bool inserted =
+        instance_dispatch_map_.insert({key, dispatch_table}).second;
+    if (inserted) {
+      instance_keys_map_[key] = instance;
+    }
+    return inserted;
   }
 
   // Removes the dispatch table and physical devices associated with |instance|.
   void RemoveInstance(VkInstance instance);
 
-  // Records that the physical device |gpu| is associated with |instance|.
-  void AddPhysicalDevice(VkInstance instance, VkPhysicalDevice gpu) {
-    absl::MutexLock lock(&gpu_instance_lock_);
-    assert((gpu_instance_map_.count(gpu) == 0 ||
-            gpu_instance_map_[gpu] == instance) &&
-           "PhysicalDevice already bound to another Instance");
-    gpu_instance_map_[gpu] = instance;
-  }
-
-  // Returns the instance associated with |gpu|, or a null handle if there is
-  // none.
-  VkInstance GetInstance(VkPhysicalDevice gpu) const {
-    absl::MutexLock lock(&gpu_instance_lock_);
-    if (auto it = gpu_instance_map_.find(gpu); it != gpu_instance_map_.end())
+  // Returns the instance associated with |instance_key|, or a null handle if
+  // there is none.
+  VkInstance GetInstance(InstanceKey instance_key) const {
+    absl::MutexLock lock(&instance_dispatch_lock_);
+    if (auto it = instance_keys_map_.find(instance_key);
+        it != instance_keys_map_.end())
       return it->second;
     return VK_NULL_HANDLE;
   }
 
-  // Records the dispatch table and get_device_proc_addr associated with
+  // Records the dispatch table and device key associated with
   // |device|.
   bool AddDevice(VkDevice device, const VkLayerDispatchTable& dispatch_table) {
+    DeviceKey key(device);
     absl::MutexLock dispatch_table_lock(&device_dispatch_lock_);
-    return device_dispatch_map_.insert({device, dispatch_table}).second;
+    const bool inserted =
+        device_dispatch_map_.insert({key, dispatch_table}).second;
+    if (inserted) {
+      device_keys_map_[key] = device;
+    }
+    return inserted;
   }
 
   // Removes the dispatch table associated with |device|.
   void RemoveDevice(VkDevice device) {
+    DeviceKey key(device);
+    absl::MutexLock lock(&device_dispatch_lock_);
+    device_dispatch_map_.erase(key);
+    device_keys_map_.erase(key);
+  }
+
+  // Returns the device associated with |device_key|, or a null handle if
+  // there is none.
+  VkDevice GetDevice(DeviceKey device_key) const {
```

Comment:
Same as `GetInstance`, can we have this take a `vkDevice` and construct the key inside?

Reply:
Same as above. It only makes sense to pass keys constructed from VkCommandBuffer and VkQueue here, passing VkDevice is an identity.

Reply:
Yep, fine to leave as it is.